esx_service/utils/auth_data.py

```diff
@@ -417,14 +420,223 @@ def remove_datastore_access_privileges(self, conn, datastore_url):
 
         return None
 
+class DBCacheManager(object):
+    """
+    Support for thread--shared in-memory RO cache.
+
+    - uses sqlite3 in -memory shared DB. One connection per thread is expected.
+    - keeps a connetion to in-mem DB to prevent it from cleaning up static cache
+    - on first enable, check we are not enabled, populate the DB
+    - gets NEW connection when asked: conn = cache.new_connection()
+    - on refresh, drains in flight Auth managers using the hshared cache,
+       drop  connection (that will deleted the DB when sqlite refcount drops to 0) and recreate cache
+    """
+
+    # this is in-memory shared DB URL
+    MEMDB_NAME = "file::memory:?cache=shared"
+
+    # - How often to refresh memdb from on-disk DB.  Default 15 min.
+    REFRESH_SEC = int(os.getenv("VMDKOPS_MEMDB_REFRESH_SEC", default=900))
+
+
+    def __init__(self):
+        # Init connection to in-memory DB.
+        # We keep one connection open to assure the DB is not garbage collected too early
+        # We refresh it every MEMDB_REFRESH_SEC i. When we refresh, we first drain all in-flight
+        # AuthernticatoinDataManagers to make sure nobody is using the copy.
+        #
+        # 'None' means the feature is disabled
+        self._memdb_con = None
+
+        # Init Barrier
+        # We need to avoid refresh (which could be scheduled by Timer on a different thread)
+        # from messing DB content for auth. mgrs in flight. So  we acquire a barrier,
+        # and wait until all auth.managers in flight land.
+        self._barrier = threading.Condition()
+        self._in_flight = 0
+
+        # Init timer for periodic refreshes
+        self._timer = None
+
+    def __del__(self):
+        pass
+
+    def _drain_and_close(self):
+        """
+        Drains Auth. managers in fight and closes last memdb connection.
+        *** EXPECTS to be run under _barrier lock.
+        """
+        logging.debug("_Drain and close")
+        # drain the in-flight users
+        self._memdb_con.close()
+        self._memdb_con = None
+        while self._in_flight > 0:
```

Comment:
I'm not sure this loop works as intended, if multiple threads can call this function. For instance, let's say that there is only 1 in flight auth mgr. Does that mean that there is only one thread that can wait on this barrier? If not, one of the threads waiting on the barrier may never be woken up. I'm not sure this is a possibility with the way the code is intended to work, but I figured I'd bring it up.